Wikipedia/Code/WMFGalleryImageListParser.m

```diff
@@ -0,0 +1,78 @@
+#import "WMFGalleryImageListParser.h"
+#import "WMFGalleryImageTag.h"
+#import "WMFGalleryImageTag+GalleryImageURL.h"
+#import <BlocksKit/BlocksKit.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface WMFGalleryImageListParser() <NSXMLParserDelegate>
+
+@property (nonatomic, strong, nullable) NSMutableArray<NSDictionary*>* parsedImgTagAttributeDicts;
+
+@end
+
+@implementation WMFGalleryImageListParser
+
+- (nullable NSArray<NSURL*>*)parseGalleryImageURLsFromHTMLString:(NSString*)HTMLString targetWidth:(NSUInteger)targetWidth;{
+
+    HTMLString = [HTMLString copy];
+
+    // NSXMLParser will update parsedImgTagAttributeDicts with a dictionary for each img tag it parses.
+    self.parsedImgTagAttributeDicts = [[NSMutableArray alloc] init];
+    
+    // First reduce HTMLString to only image tags so other funky/malformed html won't choke NSXMLParser.
+    NSString* imgTagsHtml = [self imgTagsOnlyFromHTMLString:HTMLString];
+    
+    // NSXMLParser wants a single root element.
+    imgTagsHtml = [NSString stringWithFormat:@"<base>%@</base>", imgTagsHtml];
+    
+    // Then use NSXMLParser to auto-extract tag attribute key/values to a dictionary for each image tag.
+    [self parseStringOfImgTags:imgTagsHtml];
+    
+    // Map parsedImgTagAttributeDicts to image tag model objects.
+    NSArray<WMFGalleryImageTag*>* imageTagModels = [self.parsedImgTagAttributeDicts bk_map:^id(NSDictionary* tagDict){
+        return [[WMFGalleryImageTag alloc] initWithSrc:tagDict[@"src"]
+                                                srcset:tagDict[@"srcset"]
+                                                   alt:tagDict[@"alt"]
+                                                 width:tagDict[@"width"]
+                                                height:tagDict[@"height"]
+                                         dataFileWidth:tagDict[@"data-file-width"]
+                                        dataFileHeight:tagDict[@"data-file-height"]];
+    }];    
+    
+    NSArray<NSURL*>* imageTagURLs = [[imageTagModels bk_select:^BOOL(WMFGalleryImageTag* tag){
```

Comment:
would there be value in a method that returns the WMFGalleryImageTag objects? This way the gallery could have one list that could be used to generate thumbnail and full res URLs without having to go through the parser again. The use case I'm thinking of is user taps an article image, the gallery shows a thumbnail - same res as was loaded for the article - while loading the full size image. The gallery could get the thumbnail and full size URLs from the same object.